File: app/src/main/java/com/kieran/winnipegbus/Activities/StopsMapActivity.java
package com.kieran.winnipegbus.Activities;

import android.os.Bundle;
import android.view.Menu;

import com.google.android.gms.maps.CameraUpdate;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.LatLngBounds;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;
import com.kieran.winnipegbus.R;
import com.kieran.winnipegbusbackend.FavouriteStop;
import com.kieran.winnipegbusbackend.Stop;

public class StopsMapActivity extends MapActivity {

    public static final float DEFAULT_ZOOM = 11f;
    public static final double DEFAULT_LATITUDE = 49.8954;
    public static final double DEFAULT_LONGITUDE = -97.1385;
    public static final int MAP_PADDING = 50;
    private CameraUpdate cameraUpdate;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_route_map);
        mapOnCreate();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_map, menu);
        return true;
    }

    @Override
    public void onConnected(Bundle dataBundle) {
        if(SearchResultsActivity.searchResultsList != null) {
            for (FavouriteStop favouriteStop : SearchResultsActivity.searchResultsList) {
                MarkerOptions markerOptions = new MarkerOptions();
                markerOptions.position(favouriteStop.getLatLng());
                markerOptions.title(Integer.toString(favouriteStop.getNumber()));
                markerOptions.snippet(favouriteStop.getName());
                map.addMarker(markerOptions);
            }
        }

        map.setOnInfoWindowClickListener(new GoogleMap.OnInfoWindowClickListener() {
            @Override
            public void onInfoWindowClick(Marker marker) {
                openStopTimes(new Stop(marker.getSnippet(), Integer.parseInt(marker.getTitle())));
            }
        });
        onConnected();

        moveMap();
    }

    private void moveMap() {
        double maxLat = 0;
        double maxLng = 0;
        double minLat = 0;
        double minLng = 0;

        if(cameraUpdate == null) {
            for (FavouriteStop favouriteStop : SearchResultsActivity.searchResultsList) {
                if (maxLat == 0 || favouriteStop.getLatLng().latitude > maxLat)
                    maxLat = favouriteStop.getLatLng().latitude;
                if (maxLng == 0 || favouriteStop.getLatLng().longitude > maxLng)
                    maxLng = favouriteStop.getLatLng().longitude;
                if (minLat == 0 || favouriteStop.getLatLng().latitude < minLat)
                    minLat = favouriteStop.getLatLng().latitude;
                if (minLng == 0 || favouriteStop.getLatLng().longitude < minLng)
                    minLng = favouriteStop.getLatLng().longitude;
            }

            if (maxLat != 0 && minLat != 0 && maxLng != 0 && minLng != 0) {
                LatLngBounds bounds = new LatLngBounds(new LatLng(minLat, minLng), new LatLng(maxLat, maxLng));
                cameraUpdate = CameraUpdateFactory.newLatLngBounds(bounds, MAP_PADDING);
            } else {
                cameraUpdate = CameraUpdateFactory.newLatLngZoom(new LatLng(DEFAULT_LATITUDE, DEFAULT_LONGITUDE), DEFAULT_ZOOM);
            }
            map.moveCamera(cameraUpdate);
        }
    }
}

File: app/src/main/java/com/kieran/winnipegbus/activities/StopsMapActivity.kt
package com.kieran.winnipegbus.activities

import android.os.Bundle
import android.view.Menu
import com.google.android.gms.maps.CameraUpdate
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.LatLngBounds
import com.google.android.gms.maps.model.MarkerOptions
import com.kieran.winnipegbus.R
import com.kieran.winnipegbusbackend.TransitServiceProvider
import com.kieran.winnipegbusbackend.common.FavouriteStop
import com.kieran.winnipegbusbackend.common.Stop
import com.kieran.winnipegbusbackend.interfaces.TransitService

class StopsMapActivity : MapActivity() {
    private var cameraUpdate: CameraUpdate? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_route_map)
        mapOnCreate()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_map, menu)
        return true
    }

    override fun onConnected(dataBundle: Bundle?) {
        if (SearchResultsActivity.stops != null) {
            for (favouriteStop in SearchResultsActivity.stops) {
                val markerOptions = MarkerOptions()

                markerOptions.position(if (favouriteStop.latLng != null)  LatLng(favouriteStop.latLng!!.latitude, favouriteStop.latLng!!.longitude) else null)
                markerOptions.title(favouriteStop.identifier.toString())
                markerOptions.snippet(favouriteStop.name)
                map!!.addMarker(markerOptions)
            }
        }

        map!!.isTrafficEnabled = true
        map!!.setOnInfoWindowClickListener { marker -> openStopTimes(FavouriteStop(marker.snippet, transitService.parseStringToStopIdentifier(marker.title))) }
        onConnected()

        moveMap()
    }

    private fun moveMap() {
        var maxLat = 0.0
        var maxLng = 0.0
        var minLat = 0.0
        var minLng = 0.0

        if (cameraUpdate == null) {
            for (favouriteStop in SearchResultsActivity.stops) {
                if (maxLat == 0.0 || favouriteStop.latLng!!.latitude > maxLat)
                    maxLat = favouriteStop.latLng!!.latitude
                if (maxLng == 0.0 || favouriteStop.latLng!!.longitude > maxLng)
                    maxLng = favouriteStop.latLng!!.longitude
                if (minLat == 0.0 || favouriteStop.latLng!!.latitude < minLat)
                    minLat = favouriteStop.latLng!!.latitude
                if (minLng == 0.0 || favouriteStop.latLng!!.longitude < minLng)
                    minLng = favouriteStop.latLng!!.longitude
            }

            cameraUpdate = if (maxLat != 0.0 && minLat != 0.0 && maxLng != 0.0 && minLng != 0.0) {
                val bounds = LatLngBounds(LatLng(minLat, minLng), LatLng(maxLat, maxLng))
                CameraUpdateFactory.newLatLngBounds(bounds, MAP_PADDING)
            } else {
                CameraUpdateFactory.newLatLngZoom(LatLng(DEFAULT_LATITUDE, DEFAULT_LONGITUDE), DEFAULT_ZOOM)
            }
            map!!.moveCamera(cameraUpdate!!)
        }
    }
}
